날씨정보(후) [Step1] 타코캣

Ldoy opened the pull request:
안녕하세요 찌루루~~ @jae57  많이 늦은 PR보냅니다🥲🥲🥲 노티피케이션을 줄이지 못하다 보니 흐름이 많이 복잡해졌습니다.. 너른 양해 부탁드립니다🙇‍♀️

# 고민한 점
1. 최대한 뷰컨의 역할을 분리하려 했습니다. MVC패턴을 조금 더 개선할 수 있는 방법을 찾아보았는데 기능단위로 뷰컨을 분리하고 parent-child관계를 추가해 주는 방식을 알게 되었습니다. 결과적으론 MainVC, LocationVC, TableVC 이렇게 3개의 뷰컨으로 분리를 했습니다. 
2. 네트워크 타입을 최대한 리팩토링 해 보았습니다. 위치정보와 session 정보를 전달하면 알아서 네트워크통신을 할 수 있는 타입을 구현하려 하였습니다. 결과적으론 EndPoint타입을 통해 Path, Query등을 선택하고 해당 속성을 기반으로 URL을 만든 후 URLRequest객체를 만들도록 구현 해 보았습니다. 
3. 현재 알람을 present하는 역할을 MainViewcontroller에서 하고 있습니다. 테이블뷰의 헤더뷰에 있는 버튼이라서 테이블뷰가 present하도록 해야하나 고민했는데 `테이블뷰컨이 알람을 present한다`는 것보단 메인 뷰컨에서 alert을 present한다는게 뭔가 더 어울리는 역할 같아서 현재처럼 구현 해 보았습니다. 
4. WeatherDataHolder 타입 내부에서 5일예보 날씨 정보가 가장 마지막에 들어오기 때문에 해당 case 위치에 Notification 을 구현하였습니다. 
    ```swift
    switch path {
        case .forecast :
            do {
                let decodedData = try  Parser().decode(data, to: FiveDaysForecastData.self)
                self.forcast = decodedData
                NotificationCenter.default.post(name: .reloadTableView, object: nil)
            } catch {
                print(error)
            }
            .
            .
    ```

# 질문
1. 네트워크 요청 타입에서 고쳐야 할 부분이나 개선할 부분이 있다면 무엇일지 알고싶습니다. 
2. 상황에 따라 alert의 모양이 달라질 때 찌루루는 보통 어떤 흐름으로 구현하시는지 궁금합니다. 

# 해결하지 못 한 부분
1. 기능이 추가됨에 따라 NotificationCenter를 통한 데이터 전달의 흐름을 파악하기가 힘들어졌습니다. Delegate Pattern을 시도하려했는데 PR이 너무 많이 늦어질 것 같아서 현재 리팩토링은 진행하지 못했습니다. PR을 보낸 후 적용해 보려합니다. 
2. 현재 Alert의 타입이 2가지 종류입니다. 각 AlertController마다 동일한 기능이 겹치고 있는데 이 부분 효율적으로 해결하지 못 했습니다. 
3. cell이나 headerView의 컨텐츠, 제약 등을 구현하는 View Code의 흐름이 중구난방인 것 같아 프로토콜을 통해 통일하고 싶었는데 이 부분 아직 해결하지 못했습니다. (이 글을 본 후 시도하려 합니다. [링크](https://pedroalvarez-29395.medium.com/view-code-em-ios-a-control-framework-to-eliminate-interface-builders-449fa1884958))
4. 분명 alert은 문제없이 나왔는데 PR보내면서 확인하다보니 갑자기 안되네요...다시 고쳐보겠습니다.🥲
5. `group`을 나누는 부분은 아직 깊게 고민하지 못하였습니다.

Ldoy committed:
chore: 인디케이터 속성 위치 변경

Ldoy committed:
feat: 버튼 제목 바꾸기 위해 사용자의 권한 설정여부를 알 수 있는 속성 추가

Ldoy committed:
refactor: 오픈웨더메인뷰컨이 하나의 역할만 감당하게 하기 위해 테이블뷰와 관련된 요소 위치 변경

Ldoy committed:
feat: 버튼 관리하는 버튼 컨트롤러, 테이블뷰 관리하는 컨트롤러 추가

- OpenWeatherMainViewController 이름 변경 및  alert 추가
- 위치변경 버튼 누르면 notification으로 post하도록 구현

Ldoy committed:
refactor: 테이블뷰컨에 있던 버튼컨트롤러를 해더뷰로 변경

Ldoy committed:
feat: 버튼 컨트롤러와 테이블뷰 컨트롤러 부모-자식 관계만듬

Ldoy committed:
feat: alert위치를 버튼 컨트롤러로 변경 및 locationDelegate 타입 새로운 파일로 위치변경

Ldoy committed:
chore: 커밋되지 않은 LocationManagerDelegate 타입 파일 추가

Ldoy committed:
feat: 유저가 위치 허용했을 때 뜨는 알림 UI 및 로직 구현

Ldoy committed:
refactor: 버튼컨트롤러 삭제 및 위도경도 값에 따른 alert구현

Ldoy committed:
feat: 사용하지 않는 UserState 열거형 삭제

- 역할에 따라 파일 위치 변경

Ldoy committed:
chore: 파일 위치 변경

Ldoy committed:
refactor: NotificationName 변경 및 버튼 제목 변경

jae57 commented on WeatherForecast/WeatherForecast/Extension/Notification.Name.swift:
```diff
@@ -0,0 +1,17 @@
+    static let requestLocationAgain = Notification.Name("reloadView")
+    static let showValidLocationAlert = Notification.Name("showAlert")
+    static let showInValidLocationAlert = Notification.Name("showInvalidAlert")
+    static let cleanTalbeView = Notification.Name("deleteTableViewCell")
```
Notification 이름(deleteTableViewCell)과 정적변수명(cleanTalbe오타View)이 여기만 다르네요? 의도한 건가요?

Ldoy replied:
앗 아닙니다..하핳ㅎ.. 현재는` deleteTableViewCell` 을 없애고 주소변경 실패 시 tableViewCell의 상태를 변경하는 로직을 테이블뷰에 `EmptyDelegate`를 할당해 주는 형태로 바꾸었습니다.

jae57 commented on WeatherForecast/WeatherForecast/Extension/Notification.Name.swift:
```diff
@@ -0,0 +1,17 @@
+import Foundation.NSNotification
+
+extension Notification.Name {
+    static let reloadTableView = Notification.Name("reload TableView")
```
띄어쓰기는 의도인가요?_?

Ldoy replied:
흐헉 고쳤습니다!!!🌟

jae57 commented on WeatherForecast/WeatherForecast/Service/Manager/WeatherNetworkManager.swift:
```diff
@@ -18,8 +18,6 @@ enum PathOptions {
         case data = "data"
         case forecast = "forecast"
```
rawValue 가 동일하다면 생략해줘도 됩니다!

Ldoy replied:
넵 수정하였습니다~~

jae57 commented on WeatherForecast/WeatherForecast/Service/Manager/OpenWeatherSessionDelegate.swift:
```diff
@@ -0,0 +1,35 @@
+        let requestData = proposedResponse.data
+        let pathCompnent = dataTask.currentRequest?.url?.pathComponents.last
+        
+        if let path = pathCompnent {
```
오타요!

Ldoy replied:
수정했습니다~~

jae57 commented on WeatherForecast/WeatherForecast/Service/Manager/OpenWeatherSessionDelegate.swift:
```diff
@@ -0,0 +1,35 @@
+        let pathCompnent = dataTask.currentRequest?.url?.pathComponents.last
+        
+        if let path = pathCompnent {
+            WeatherDataHolder.shared.generate(path, requestData)
```
Swift API Design Guidelines 의 Argument Labels 에 명시된 특수한 케이스를 제외하고는 
라벨을 명시하는 것을 권장하고 있습니다.
https://swift.org/documentation/api-design-guidelines/#argument-labels

Ldoy replied:
넵 수정했습니다~!

jae57 commented on WeatherForecast/WeatherForecast/Service/Manager/OpenWeatherSessionDelegate.swift:
```diff
@@ -0,0 +1,35 @@
+        }
+    }
+    
+    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, willCacheResponse proposedResponse: CachedURLResponse, completionHandler: @escaping (CachedURLResponse?) -> Void) {
```
`func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {}` 가 아닌 해당 메서드를 이용한 이유는 무엇인가요?

Ldoy replied:
내부에 data 속성을 선언하여 `append`하지 않아도 되고 바로 데이터를 사용할 수 있어서 위 메서드를 사용했습니다~!

jae57 commented on WeatherForecast/WeatherForecast/View/TableView/OpenWeatherHeaderView.swift:
```diff
@@ -61,7 +99,6 @@ class OpenWeatherHeaderView: UITableViewHeaderFooterView {
         let convertedmaxTem = TemperatureConverter(celciusTemperature: maxTem).convertedTemperature
```
카멜 케이스가 제대로 적용되지 않은 것 같네요! 
Tem 과 같이 줄임말은 오해를 일으킬 수 있습니다. 되도록이면 풀어써주세요!
![image](https://user-images.githubusercontent.com/15043840/138336184-71120b65-2c67-49d9-ba93-8ff08207a925.png)
https://swift.org/documentation/api-design-guidelines/#use-terminology-well

Ldoy replied:
넵 수정하였습니다~~~!

jae57 commented on WeatherForecast/WeatherForecast/Controller/LocationViewController.swift:
```diff
@@ -0,0 +1,35 @@
+    }
+    
+    @objc func requestLocationAgain() {
+        self.locationManager.requestLocation()
```
self에 대한 일관된 규칙을 가지고가면 좋겠어요!

Ldoy replied:
아하 넵!! 추후엔 속성 및 메서드에 모두 self를 붙이는 쪽으로 규칙을 가지고 진행하겠습니다~!

jae57 commented on WeatherForecast/WeatherForecast/Controller/TableViewController.swift:
```diff
@@ -0,0 +1,102 @@
+    }
+    
+    @objc private func reloadTableView() {
+        DispatchQueue.main.async { [self] in
```
[self] 는 어떤 의도로 작성된 것일까요?

Ldoy replied:
앗 weak self고민하다가 작성한건데 지운다는게 놓쳤습니다~! 수정했습니다~!

jae57 commented on WeatherForecast/WeatherForecast/Controller/TableViewController.swift:
```diff
@@ -0,0 +1,102 @@
+    @objc private func reloadTableView() {
+        DispatchQueue.main.async { [self] in
+            self.tableView.dataSource = self.tableViewDataSource
+            self.tableView.delegate = self.tableViewDelegate
```
테이블뷰를 reload 해줄 때마다 dataSource 와 delegate 를 할당해주는 이유는 무엇인가요?

Ldoy replied:
현재 위치 요청 후 CLGeocoder를 통해 주소를 얻지 못하면 `EmptyDelegate`, `EmptyDataSource` 로 delegate 와 datasource를 바꿔주고 있어서 reload할 때마다 새로 할당 해 주었습니다~!

jae57 reviewed (commented):
@Ldoy 
🙋‍♂️ 안녕하세요~~~ 타코캣 !

고민한 점과 질문 잘 읽어보았습니다!

**고민1** - 뷰컨을 분리하는 부분은 정답이 없는 부분이지만 제 생각을 말씀드리자면
뷰컨을 쪼갬으로써 각 뷰컨들의 역할이 명확해진다는 장점도 있지만
containerView를 구성하는것과 컨트롤러를 관리하는 컨트롤러도 생기고 조금은 복잡해진다는 단점도 존재하는것 같습니다.
너무 사이즈가 크거나 재사용이 빈번하게 일어날경우에는 분리를 해도 좋을것 같지만
간단한 경우에는 extension정도로만 분리를 해도 괜찮지 않을까 생각해봅니다.

**고민2, 질문1** - 저는 이럴때 사람들이 많이 사용하는 구조를 많이 참고해보면서 왜 이렇게 짯을까도 고민을 해보고, 
직접 적용해보기도 합니다. 그래서 네트워크 유명한 라이브러리는 moya가 있는데 한번 참고해보시면 어떨까요??

**고민3** - 어떤 뷰컨이든 상관은 없을것 같습니다. 
그래도 조금 고민을 해보자면 테이블뷰의 버튼을 눌렀을때 동작하기때문에 
저는 테이블뷰컨에서 처리하는것이 맞지않을까 싶기도 합니다. 
왜냐하면 그러면 테이블 뷰컨을 다른곳에서 재사용할때 더 편하게 사용할수 있지 않을까 생각했습니다.

보통 UIViewController extension으로 presentAlert와 같은 형태로 미리 구현을 해두는 편입니다.
개인적으로 https://github.com/cuzv/ExtensionKit/blob/master/Sources/Extension/UIViewController%2BExtension.swift 이런 형태를 많이 사용하는편 입니다.

추가적으로 몇가지 사용성 면에서 개선되면 좋을만한 것들을 적어봅니다!
**1**
실행해본 화면입니다.
![image](https://user-images.githubusercontent.com/15043840/138337380-7d53d62a-147c-4002-b778-f13d2fef006a.png)
기온값이 날씨 이미지에 가려지고 있습니다 😢

**2**
위도, 경도 입력하는 부분에 한글을 입력했을 때 입니다. 
입력을 막거나 사용자에게 어떤 상황인지 알리는 것이 필요해보이네요!
<img src="https://user-images.githubusercontent.com/15043840/138338380-9b299240-7ea2-4b95-b6d9-026b1c0a87fc.png" width="300">

**3**
위치설정 버튼을 눌렀을 때 뜨는 얼럿을 종료할 방법이 없네요!
<img src="https://user-images.githubusercontent.com/15043840/138490420-5d619709-a4d0-4c98-b6f1-025479e02a7f.png" width="300">

Ldoy committed:
feat: alert 구현 완료 및 오타 수정

Ldoy committed:
refactor: 버튼 alert로직을 Notification에서 delegate pattern으로 변경

Ldoy committed:
feat: 앱 배경화면 구현

Ldoy committed:
refactor: 위치 잘못 입력했을 때 테이블뷰에게 알려주는 로직 델리게이트로 변경

Ldoy committed:
refactor: 자주사용하는 메서드 전역함수로 변경

Ldoy committed:
refactor: 현재위치가 alert의 위도, 경도 나타나도록 로직 수정

- LocationDelegate에서 해당 위치가 주소로 변경될 수 있는 경우 locationManager의 속성에 저장하도록 함

Ldoy committed:
feat: request메소드의 argument label추가

Ldoy committed:
feat: WeatherDataHolder의 generate메서드 파라미터 이름 수정

Ldoy committed:
refactor: OpenWeatherHeaderView 메서드 내부 변수명 수정

Ldoy commented:
1. 오토레이아웃이 안잡히는 문제를 해결 해 보았습니다. 
- 11pro
<img src = "https://user-images.githubusercontent.com/34529917/139079474-ce8361e8-b194-4da0-b8c3-06f660474af3.png" width = 200, height = 400>
- 12pro
<img src = "https://user-images.githubusercontent.com/34529917/139079490-b729ec74-a28a-4e85-94c1-9c6b9bdc0d28.png" width = 200, height = 400>


2. 위치정보를 알 수 없을 때 아래와 같은 화면이 나오도록 구현 해 보았습니다. 
<img src = "https://user-images.githubusercontent.com/34529917/139079762-01fabdf7-230d-4026-a379-e12b0d6eafbf.png" width = 200, height = 400>

많이 늦은 코멘트 반영 죄송합니다🙇‍♀️ 찌루루 @jae57  지난 4주간 감사했습니다~~🤩
